### DISCUSSION TITLE
Rekey after entropy exhaustion

### DISCUSSION THREAD

**Discussion by amdrsantos** [strongswan/strongswan#1556] [Q&A] [answered]
I am running some tests with strongSwan in harsh scenarios, using stongSwan v5.9.9 and OpenSSL 3.0.x crypto library for all ciphers.

One of those scenarios is when the entropy generator becomes unavailable, there is no entropy to generate a new key.
In that case, all the applications that use OpenSSL hang when looking for entropy, like for example:

```
root@machine# strace ssh-keygen -t rsa -b 4096 -N '' -f /tmp/any.key
...
socket(AF_UNIX, SOCK_STREAM, 0)         = 4
fcntl(4, F_GETFL)                       = 0x2 (flags O_RDWR)
connect(4, {sa_family=AF_UNIX, sun_path="/var/run/egd-pool"}, 19) = 0
write(4, "\1\20", 2)                    = 2
read(4, `
```

In case there is a rekey in one Child SA or IKE SA, strongSwan also hangs (once is using OpenSSL).
```
Feb 23 12:12:08 machine charon[105277]: 09[CFG] vici rekey IKE_SA 'CentOs'
```

```
[pid 34230] socket(AF_UNIX, SOCK_STREAM, 0) = 19<UNIX:[1359556]>
[pid 34230] fcntl(19<UNIX:[1359556]>, F_GETFL) = 0x2 (flags O_RDWR)
[pid 34230] connect(19<UNIX:[1359556]>, {sa_family=AF_UNIX, sun_path="/var/run/egd-pool"}, 19) = 0
[pid 34230] write(19<UNIX:[1359556]>, "\1\20", 2) = 2
[pid 34230] read(19<UNIX:[1359556]>,  <unfinished ...>
```

But in this case, when the rekey hangs, there is no event the connection is down.
Is this expected behavior?

**Comment by tobiasbrunner** [comment, marked as answer]
> But in this case, when the rekey hangs, there is no event the connection is down.

How is it down when the rekeying hangs?

> Is this expected behavior?

What else would you expect exactly?

**Reply by amdrsantos** [reply]
> How is it down when the rekeying hangs?

The peer sends the connection down after some time
The local strongSwan only sends the connection down on:
```
Feb 24 16:03:59 machine charon[198881]: 03[KNL] received a XFRM_MSG_EXPIRE
Feb 24 16:03:59 machine  charon[198881]: 03[KNL] creating delete job for CHILD_SA ESP/0xc8811207/10.220.196.10
```
Still, no event is triggered.


> What else would you expect exactly?

I would expect, at least, to receive an event when the connection is down

**Reply by tobiasbrunner** [reply]
> The peer sends the connection down after some time

So the peer does not initiate a rekeying of the SA? (Not that a rekey request from the peer would make a difference if there is no entropy.)

> The local strongSwan only sends the connection down on:

I see, yeah it definitely expires after some time (depends on your [lifetime/rekey settings](https://docs.strongswan.org/docs/5.9/config/rekeying.html)). But it's not yet down when the rekeying hangs just for a short while.

> Still, no event is triggered.

What kind of event are you referring to? ike-updown/child-updown? That won't be possible as the delete job will not be able to check out the IKE_SA of the expired IKE/CHILD_SA as the other thread that is trying to initiate the rekeying is waiting for entropy for the nonce or DH key and has it checked out. If that happens multiple times, all available threads of the thread pool will eventually be blocked and no jobs can be handled anymore.

**Reply by amdrsantos** [reply]
> What kind of event are you referring to? ike-updown/child-updown?

Yes.

> That won't be possible as the delete job will not be able to check out the IKE_SA of the expired IKE/CHILD_SA as the other thread that is trying to initiate the rekeying is waiting for entropy for the nonce or DH key and has it checked out.

So this is done by design. Do consider changing it?

**Reply by tobiasbrunner** [reply]
Well, what alternatives are there? We call e.g. an OpenSSL function that is expected to generate a DH key, if that blocks, what can we do?
